url3aa01d34ec_katsuster_ememu_tgz-pJ8-net_katsuster_ememu_ui_MainJ8/src/net/katsuster/ememu/arm/core/ExecStageThumb.java

```diff
@@ -0,0 +1,2318 @@
+            printDisasm(inst, "strb",
+                    String.format("%s, [%s, %s]",
+                            getRegName(rd), getRegName(rn), getRegName(rm)));
+            return;
+        }
+
+        vaddr = getReg(rn) + getReg(rm);
+
+        paddr = getMMU().translate(vaddr, 1, false, getCPSR().isPrivMode(), false);
+        if (getMMU().isFault()) {
+            getMMU().clearFault();
+            return;
+        }
+
+        if (!tryWrite_a32(paddr, 1)) {
+            raiseException(ARMv5.EXCEPT_ABT_DATA,
+                    String.format("strb [%08x]", paddr));
+            return;
+        }
+        write8_a32(paddr, (byte) getReg(rd));
+    }
+
+    /**
+     * ハーフワードストア命令。
+     *
+     * @param inst Thumb 命令
+     * @param exec デコードと実行なら true、デコードのみなら false
+     */
+    public void executeStrh1(InstructionThumb inst, boolean exec) {
+        int imm5_raw = inst.getField(6, 5);
+        int rn = inst.getField(3, 3);
+        int rd = inst.getRdField();
+        int imm5 = imm5_raw << 1;
+        int vaddr, paddr;
+
+        if (!exec) {
+            printDisasm(inst, "strh",
+                    String.format("%s, [%s, #%d]",
+                            getRegName(rd),
+                            getRegName(rn), imm5));
+            return;
+        }
+
+        vaddr = getReg(rn) + imm5;
+
+        paddr = getMMU().translate(vaddr, 2, false, getCPSR().isPrivMode(), false);
+        if (getMMU().isFault()) {
+            getMMU().clearFault();
+            return;
+        }
+
+        if (!tryWrite_a32(paddr, 2)) {
+            raiseException(ARMv5.EXCEPT_ABT_DATA,
+                    String.format("strh [%08x]", paddr));
+            return;
+        }
+        write16_a32(paddr, (short) getReg(rd));
+    }
+
+
+    /**
+     * ハーフワードストア命令。
+     *
+     * @param inst Thumb 命令
+     * @param exec デコードと実行なら true、デコードのみなら false
+     */
+    public void executeStrh2(InstructionThumb inst, boolean exec) {
+        int rm = inst.getField(6, 3);
+        int rn = inst.getField(3, 3);
+        int rd = inst.getRdField();
+        int vaddr, paddr;
+
+        if (!exec) {
+            printDisasm(inst, "strh",
+                    String.format("%s, [%s, %s]",
+                            getRegName(rd), getRegName(rn), getRegName(rm)));
+            return;
+        }
+
+        vaddr = getReg(rn) + getReg(rm);
+
+        paddr = getMMU().translate(vaddr, 2, false, getCPSR().isPrivMode(), false);
+        if (getMMU().isFault()) {
+            getMMU().clearFault();
+            return;
+        }
+
+        if (!tryWrite_a32(paddr, 2)) {
+            raiseException(ARMv5.EXCEPT_ABT_DATA,
+                    String.format("strh [%08x]", paddr));
+            return;
+        }
+        write16_a32(paddr, (short) getReg(rd));
+    }
+
+    /**
+     * プッシュ命令。
+     *
+     * @param inst Thumb 命令
+     * @param exec デコードと実行なら true、デコードのみなら false
+     */
+    public void executePush(InstructionThumb inst, boolean exec) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 13. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 1816-1832 into a separate method.